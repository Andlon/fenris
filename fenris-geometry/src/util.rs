use nalgebra::{RealField, UnitVector3, Vector3};

pub fn compute_orthonormal_vectors_3d<T: RealField>(vector: &UnitVector3<T>) -> [UnitVector3<T>; 2] {
    // Ported from
    // https://github.com/dimforge/parry/blob/ac8dcf0197066cd2413a20e4420961b4694996c0/src/utils/wops.rs#L120-L138
    // originally based on the Pixar paper "Building an Orthonormal Basis, Revisited",
    // https://graphics.pixar.com/library/OrthonormalB/paper.pdf
    let v = vector;
    let sign = T::copysign(T::one(), v.z);
    let a = -T::one() / (sign + v.z);
    let b = v.x * v.y * a;

    [
        Vector3::new(T::one() + sign * v.x * v.x * a, sign * b, -sign * v.x),
        Vector3::new(b, sign + v.y * v.y * a, -v.y),
    ]
    .map(UnitVector3::new_unchecked)
}

/// Compares two arrays for *shift-invariant* equality with the given comparator function.
///
/// If `X` and `Y` are the two arrays, then the two arrays are shift-invariant equal if X can be shifted/rotated
/// by some constant `n` such that `Shift(X) == Y`.
pub fn slices_are_equal_shift_invariant<T, C: Fn(&T, &T) -> bool>(x: &[T], y: &[T], comparator: C) -> bool {
    let n = x.len();
    if y.len() != n {
        return false;
    } else if n == 0 {
        // Empty arrays are always equal
        return true;
    }

    'outer_loop: for i_start in 0..n {
        for (j, y_j) in y.iter().enumerate() {
            let x_j_shifted = &x[(j + i_start) % n];
            if !comparator(x_j_shifted, y_j) {
                continue 'outer_loop;
            }
        }
        return true;
    }

    return false;
}

// Base macro for line segment assertions
#[doc(hidden)]
#[macro_export]
macro_rules! assert_line_segments_approx_equal_base {
    ($msg_handler:expr, $segment1:expr, $segment2:expr, abstol = $tol:expr) => {{
        use fenris_geometry::util::slices_are_equal_shift_invariant;
        use matrixcompare::comparators::AbsoluteElementwiseComparator;
        use matrixcompare::compare_matrices;
        use nalgebra::Point2;

        let tol = $tol.clone();
        // Type check: Makes for an easier error message than failing specific methods
        let (segment1, segment2): (&LineSegment2d<_>, &LineSegment2d<_>) = (&$segment1, &$segment2);
        let vertices1 = [segment1.start().clone(), segment1.end().clone()];
        let vertices2 = [segment2.start().clone(), segment2.end().clone()];
        let comparator = AbsoluteElementwiseComparator { tol };
        let comparator = |a: &Point2<f64>, b: &Point2<f64>| compare_matrices(&a.coords, &b.coords, &comparator).is_ok();
        let vertices_are_shift_invariant_equal = slices_are_equal_shift_invariant(&vertices1, &vertices2, comparator);
        if !vertices_are_shift_invariant_equal {
            let msg = format!(
                "Line segments are not (approximately) equal to absolute tolerance {tol}.
Segment1: {:?}
Segment2: {:?}",
                segment1, segment2
            );

            return $msg_handler(msg);
        }
    }};
}

#[macro_export]
macro_rules! assert_line_segments_approx_equal {
    ($segment1:expr, $segment2:expr, abstol = $tol:expr) => {{
        let msg_handler = |msg| panic!("{}", msg);
        $crate::assert_line_segments_approx_equal_base!(msg_handler, $segment1, $segment2, abstol = $tol);
    }};
}
