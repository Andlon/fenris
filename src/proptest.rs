use crate::element::{Tet4Element, Tri3d2Element};
use crate::geometry::proptest::Triangle3dParams;
use crate::geometry::Orientation::Counterclockwise;
use crate::mesh::procedural::create_rectangular_uniform_quad_mesh_2d;
use crate::mesh::QuadMesh2d;
use ::proptest::prelude::*;
use fenris_geometry::proptest::Triangle2dParams;
use fenris_geometry::{Triangle2d, Triangle3d};
use nalgebra::{Point2, Point3, Vector2};
use std::cmp::max;

pub fn point2() -> impl Strategy<Value = Point2<f64>> {
    // Pick a reasonably small range to pick coordinates from,
    // otherwise we can easily get floating point numbers that are
    // so ridiculously large as to break anything we might want to do with them
    let range = -10.0..10.0;
    [range.clone(), range.clone()].prop_map(|[x, y]| Point2::new(x, y))
}

pub fn point3() -> impl Strategy<Value = Point3<f64>> {
    // Pick a reasonably small range to pick coordinates from,
    // otherwise we can easily get floating point numbers that are
    // so ridiculously large as to break anything we might want to do with them
    let range = -10.0..10.0;
    [range.clone(), range.clone(), range.clone()].prop_map(|[x, y, z]| Point3::new(x, y, z))
}

impl Arbitrary for Tri3d2Element<f64> {
    type Parameters = ();
    type Strategy = BoxedStrategy<Self>;

    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
        any_with::<Triangle2d<f64>>(Triangle2dParams::default().with_orientation(Counterclockwise))
            .prop_map(|triangle| Self::from(triangle))
            .boxed()
    }
}

impl Arbitrary for Tet4Element<f64> {
    // TODO: Reasonable parameters?
    type Parameters = ();
    type Strategy = BoxedStrategy<Self>;

    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
        any_with::<Triangle3d<f64>>(Triangle3dParams::default().with_orientation(Counterclockwise))
            .prop_flat_map(|triangle| {
                // To create an arbitrary tetrahedron element, we take a counter-clockwise oriented
                // triangle, and pick a point somewhere on the "positive" side of the
                // triangle plane. We do this by associating a parameter with each
                // tangent vector defined by the sides of the triangle,
                // plus a non-negative parameter that scales along the normal direction
                let range = -10.0..10.0;
                let tangent_params = [range.clone(), range.clone(), range.clone()];
                let normal_param = 0.0..=10.0;
                (Just(triangle), tangent_params, normal_param)
            })
            .prop_map(|(triangle, tangent_params, normal_param)| {
                let mut tangent_pos = triangle.centroid();

                for (side, param) in triangle.sides().iter().zip(&tangent_params) {
                    tangent_pos.coords += *param * side;
                }
                let coord = tangent_pos + normal_param * triangle.normal_dir().normalize();
                Tet4Element::from_vertices([triangle.0[0], triangle.0[1], triangle.0[2], coord])
            })
            .boxed()
    }
}

// Returns a strategy in which each value is a triplet (cells_per_unit, units_x, units_y)
// such that cells_per_unit^2 * units_x * units_y <= max_cells
fn rectangular_uniform_mesh_cell_distribution_strategy(
    max_cells: usize,
) -> impl Strategy<Value = (usize, usize, usize)> {
    let max_cells_per_unit = f64::floor(f64::sqrt(max_cells as f64)) as usize;
    (1..=max(1, max_cells_per_unit))
        .prop_flat_map(move |cells_per_unit| (Just(cells_per_unit), 0..=max_cells / (cells_per_unit * cells_per_unit)))
        .prop_flat_map(move |(cells_per_unit, units_x)| {
            let units_y_strategy = 0..=max_cells / (cells_per_unit * cells_per_unit * max(1, units_x));
            (Just(cells_per_unit), Just(units_x), units_y_strategy)
        })
}

pub fn rectangular_uniform_mesh_strategy(unit_length: f64, max_cells: usize) -> impl Strategy<Value = QuadMesh2d<f64>> {
    rectangular_uniform_mesh_cell_distribution_strategy(max_cells).prop_map(
        move |(cells_per_unit, units_x, units_y)| {
            create_rectangular_uniform_quad_mesh_2d(
                unit_length,
                units_x,
                units_y,
                cells_per_unit,
                &Vector2::new(0.0, 0.0),
            )
        },
    )
}

#[cfg(test)]
mod tests {
    use super::rectangular_uniform_mesh_cell_distribution_strategy;
    use crate::geometry::proptest::{
        convex_quad2d_strategy_f64, nondegenerate_convex_quad2d_strategy_f64, nondegenerate_triangle2d_strategy_f64,
    };
    use crate::geometry::Orientation;
    use proptest::prelude::*;

    proptest! {
        #[test]
        fn rectangular_uniform_mesh_cell_distribution_strategy_respects_max_cells(
            (max_cells, cells_per_unit, units_x, units_y)
             in (0..20usize).prop_flat_map(|max_cells| {
                rectangular_uniform_mesh_cell_distribution_strategy(max_cells)
                    .prop_map(move |(cells_per_unit, units_x, units_y)| {
                    (max_cells, cells_per_unit, units_x, units_y)
                })
             })
        ) {
            // Test that the distribution strategy for rectangular meshes
            // respects the maximum number of cells given
            prop_assert!(cells_per_unit * cells_per_unit * units_x * units_y <= max_cells);
        }

        #[test]
        fn convex_quads_are_convex(quad in convex_quad2d_strategy_f64()) {
            prop_assert!(quad.concave_corner().is_none());
        }

        #[test]
        fn nondegenerate_triangles_have_positive_area(
            triangle in nondegenerate_triangle2d_strategy_f64()
        ){
            prop_assert!(triangle.area() > 0.0);
            prop_assert!(triangle.orientation() == Orientation::Counterclockwise);
        }

        #[test]
        fn nondegenerate_quads_have_positive_area(
            quad in nondegenerate_convex_quad2d_strategy_f64()
        ){
            prop_assert!(quad.area() > 0.0);
        }
    }
}
