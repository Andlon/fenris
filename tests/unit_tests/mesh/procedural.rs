use fenris::assembly::global::assemble_scalar;
use fenris::connectivity::Connectivity;
use fenris::element::{ElementConnectivity, FiniteElement, SurfaceFiniteElement};
use fenris::integrate::{dependency::NoDeps, FnFunction, UFunction};
use fenris::integrate::{integrate_over_element, volume_form, ElementIntegralAssemblerBuilder};
use fenris::io::vtk::FiniteElementMeshDataSetBuilder;
use fenris::mesh::procedural::{create_rectangular_uniform_hex_mesh, create_rectangular_uniform_tet_mesh};
use fenris::quadrature::CanonicalMassQuadrature;
use fenris::quadrature::Quadrature;
use fenris::util::global_vector_from_point_fn;
use fenris_geometry::AxisAlignedBoundingBox3d;
use matrixcompare::prop_assert_scalar_eq;
use nalgebra::coordinates::XYZ;
use nalgebra::{dvector, vector, Point3, Vector1, Vector3, Vector4, U1};
use proptest::prelude::*;
use std::path::PathBuf;

#[test]
fn rectangular_uniform_tet_mesh_basics() {
    let base_path = PathBuf::from("data/unit_tests/mesh/procedural");
    for res in [1, 2] {
        let mesh = create_rectangular_uniform_tet_mesh(1.0, 1, 1, 1, res);
        FiniteElementMeshDataSetBuilder::from_mesh(&mesh)
            .try_export(base_path.join(format!("basic_uniform_{res}.vtk")))
            .unwrap();
        // The idea here is not to inspect the actual debug output but instead to
        // look at the VTK files and check that they look reasonable
        insta::assert_debug_snapshot!(format!("mesh_{res}"), mesh);
    }
}

fn empty_tet_mesh_params() -> impl Strategy<Value = [usize; 4]> {
    let strategy = prop_oneof![Just(0), 0usize..3];
    [strategy.clone(), strategy.clone(), strategy.clone(), strategy]
        .prop_filter("At least one of the parameters must be zero", |params: &[usize; 4]| {
            params.iter().product::<usize>() == 0
        })
}

#[derive(Debug)]
struct RectangularUniformTetMeshParams {
    unit_length: f64,
    units: [usize; 3],
    resolution: usize,
}

impl Arbitrary for RectangularUniformTetMeshParams {
    type Parameters = ();

    fn arbitrary_with(_args: ()) -> Self::Strategy {
        let units = [1usize..2, 1..2, 1..2];
        (0.1f64..10.0, units, 1..3usize)
            .prop_map(|(unit_length, units, resolution)| Self {
                unit_length,
                units,
                resolution,
            })
            .boxed()
    }

    type Strategy = BoxedStrategy<Self>;
}

proptest! {
    #[test]
    fn rectangular_uniform_tet_mesh_empty([units_x, units_y, units_z, cells_per_unit] in empty_tet_mesh_params()) {
        let mesh = create_rectangular_uniform_tet_mesh(1.0, units_x, units_y, units_z, cells_per_unit);
        prop_assert!(mesh.vertices().is_empty());
        prop_assert!(mesh.connectivity().is_empty());
    }

    #[test]
    fn rectangular_uniform_tet_mesh_polynomial_integral(params: RectangularUniformTetMeshParams) {
        // Integrate a function f(x, u), where u = u(x) is a linear function
        // (so that both linear tets and trilinear hex elements are able to represent it exactly)
        // and f overall is linear.
        let u = |p: &Point3<_>| vector![4.0 * p.x + 2.0 * p.y - 2.0 * p.z + 3.0];
        let f = FnFunction::new(|p: &Point3<_>, u: &Vector1<_>| vector![-3.0 * p.x + 5.0 * p.y - 2.0 * p.z + 5.0 - 3.0 * u[0]]);

        let RectangularUniformTetMeshParams { unit_length, units, resolution } = params;
        let [nx, ny, nz] = units;
        let tet_mesh = create_rectangular_uniform_tet_mesh(unit_length, nx, ny, nz, resolution);
        let tet_quadrature = tet_mesh.canonical_mass_quadrature();
        let u_tet = global_vector_from_point_fn(tet_mesh.vertices(), u);

        // Use minimal element hex mesh to compute reference value on
        let hex_mesh = create_rectangular_uniform_hex_mesh(unit_length, nx, ny, nz, 1);
        let hex_quadrature = hex_mesh.canonical_mass_quadrature();
        let u_hex = global_vector_from_point_fn(hex_mesh.vertices(), u);

        let hex_assembler = ElementIntegralAssemblerBuilder::new()
            .with_quadrature_table(&hex_quadrature)
            .with_space(&hex_mesh)
            .with_integrand(f.clone())
            .with_interpolation_weights(&u_hex)
            .build_integrator();
        let hex_result = assemble_scalar(&hex_assembler).unwrap();

        let tet_assembler = ElementIntegralAssemblerBuilder::new()
            .with_quadrature_table(&tet_quadrature)
            .with_space(&tet_mesh)
            .with_integrand(f)
            .with_interpolation_weights(&u_tet)
            .build_integrator();
        let tet_result = assemble_scalar(&tet_assembler).unwrap();

        prop_assert_scalar_eq!(tet_result, hex_result, comp = abs, tol = hex_result.abs() * 1e-12);
    }

    #[test]
    fn rectangular_uniform_tet_mesh_geometric_invariants(params: RectangularUniformTetMeshParams) {
        let RectangularUniformTetMeshParams { unit_length, units, resolution } = params;
        let [nx, ny, nz] = units;
        let tet_mesh = create_rectangular_uniform_tet_mesh(unit_length, nx, ny, nz, resolution);

        let extents = [nx, ny, nz].map(|n| (n as f64) * unit_length);
        let aabb = AxisAlignedBoundingBox3d::from_points(tet_mesh.vertices()).unwrap();
        prop_assert_eq!(aabb.min(), &Point3::origin());
        prop_assert_eq!(aabb.max(), &Point3::from(extents));

        for connectivity in tet_mesh.connectivity() {
            let volume_element = connectivity.element(tet_mesh.vertices()).unwrap();
            // Jacobian is constant so can pick any reference point
            let j_det = volume_element.reference_jacobian(&Point3::origin()).determinant();
            prop_assert!(j_det > 0.0, "element is inverted");
        }
    }

    #[test]
    fn rectangular_uniform_tet_mesh_element_wise_divergence(params: RectangularUniformTetMeshParams) {
        // We check that the divergence theorem applies to each element. This implicitly
        // is intended to verify that all elements have correct orientation. This is however
        // somewhat redundant given that checking for inversion already accomplishes the same thing.
        // However it acts as an additional sanity check that the implementations for the
        // normal computation and face connectivity are sane

        let RectangularUniformTetMeshParams { unit_length, units, resolution } = params;
        let [nx, ny, nz] = units;
        let tet_mesh = create_rectangular_uniform_tet_mesh(unit_length, nx, ny, nz, resolution);
        let tet_quadrature = fenris::quadrature::total_order::tetrahedron(1).unwrap();
        let tri_quadrature = fenris::quadrature::total_order::triangle(2).unwrap();

        let div_f = |p: &Point3<f64>| {
            let XYZ { x, y, z } = *p.coords;
            vector![
                (6.0 * x + 2.0)
                + (4.0 * y + 3.0 - 3.0 * x)
                + (2.0 * z - 3.0 - 3.0 * x)
            ]
        };
        let div_f = FnFunction::new(div_f)
            .with_dependencies::<NoDeps<U1>>();

        let f = |p: &Point3<f64>| {
            let XYZ { x, y, z } = *p.coords;
            vector![
                3.0 * x.powi(2) + 2.0 * x + 4.0 * y * z - 3.0,
                2.0 * y.powi(2) + 3.0 * y - 3.0 * x * y + 2.0,
                z.powi(2) - 3.0 * z - 3.0 * x * z + x * y - 4.0
            ]
        };

        for connectivity in tet_mesh.connectivity() {
            let volume_element = connectivity.element(tet_mesh.vertices()).unwrap();
            // Interpolation weights are irrelevant here since our function does not depend on u,
            // just use zero weights.
            let u = Vector4::repeat(0.0);
            let div_integral = integrate_over_element(
                &div_f,
                &volume_element,
                &tet_quadrature,
                &u,
                &mut Default::default())[0];

            let mut flux_integral = 0.0;
            for face_idx in 0 .. connectivity.num_faces() {
                let face_conn = connectivity.get_face_connectivity(face_idx).unwrap();
                let face_element = face_conn.element(tet_mesh.vertices()).unwrap();

                for (w, xi, _) in tri_quadrature.iter() {
                    let n = face_element.normal(xi);
                    let x = face_element.map_reference_coords(xi);
                    let j = face_element.reference_jacobian(xi);
                    flux_integral += w * volume_form(&j) * n.dot(&f(&x));
                }
            }
            let tol = 1e-12 * div_integral.abs();
            prop_assert_scalar_eq!(flux_integral, div_integral, comp = abs, tol = tol);
        }
    }
}
