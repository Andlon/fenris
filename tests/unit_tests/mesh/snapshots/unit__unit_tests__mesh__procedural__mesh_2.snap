---
source: tests/unit_tests/mesh/procedural.rs
expression: mesh
---
Mesh {
    vertices: [
        [
            0.0,
            0.0,
            0.0,
        ],
        [
            0.5,
            0.0,
            0.0,
        ],
        [
            1.0,
            0.0,
            0.0,
        ],
        [
            0.0,
            0.5,
            0.0,
        ],
        [
            0.5,
            0.5,
            0.0,
        ],
        [
            1.0,
            0.5,
            0.0,
        ],
        [
            0.0,
            1.0,
            0.0,
        ],
        [
            0.5,
            1.0,
            0.0,
        ],
        [
            1.0,
            1.0,
            0.0,
        ],
        [
            0.0,
            0.0,
            0.5,
        ],
        [
            0.5,
            0.0,
            0.5,
        ],
        [
            1.0,
            0.0,
            0.5,
        ],
        [
            0.0,
            0.5,
            0.5,
        ],
        [
            0.5,
            0.5,
            0.5,
        ],
        [
            1.0,
            0.5,
            0.5,
        ],
        [
            0.0,
            1.0,
            0.5,
        ],
        [
            0.5,
            1.0,
            0.5,
        ],
        [
            1.0,
            1.0,
            0.5,
        ],
        [
            0.0,
            0.0,
            1.0,
        ],
        [
            0.5,
            0.0,
            1.0,
        ],
        [
            1.0,
            0.0,
            1.0,
        ],
        [
            0.0,
            0.5,
            1.0,
        ],
        [
            0.5,
            0.5,
            1.0,
        ],
        [
            1.0,
            0.5,
            1.0,
        ],
        [
            0.0,
            1.0,
            1.0,
        ],
        [
            0.5,
            1.0,
            1.0,
        ],
        [
            1.0,
            1.0,
            1.0,
        ],
        [
            0.25,
            0.25,
            0.25,
        ],
        [
            0.75,
            0.25,
            0.25,
        ],
        [
            0.25,
            0.75,
            0.25,
        ],
        [
            0.75,
            0.75,
            0.25,
        ],
        [
            0.25,
            0.25,
            0.75,
        ],
        [
            0.75,
            0.25,
            0.75,
        ],
        [
            0.25,
            0.75,
            0.75,
        ],
        [
            0.75,
            0.75,
            0.75,
        ],
    ],
    connectivity: [
        Tet4Connectivity(
            [
                27,
                28,
                13,
                10,
            ],
        ),
        Tet4Connectivity(
            [
                27,
                28,
                4,
                13,
            ],
        ),
        Tet4Connectivity(
            [
                27,
                28,
                1,
                4,
            ],
        ),
        Tet4Connectivity(
            [
                27,
                28,
                10,
                1,
            ],
        ),
        Tet4Connectivity(
            [
                0,
                3,
                12,
                27,
            ],
        ),
        Tet4Connectivity(
            [
                0,
                12,
                9,
                27,
            ],
        ),
        Tet4Connectivity(
            [
                27,
                29,
                4,
                3,
            ],
        ),
        Tet4Connectivity(
            [
                27,
                29,
                13,
                4,
            ],
        ),
        Tet4Connectivity(
            [
                27,
                29,
                12,
                13,
            ],
        ),
        Tet4Connectivity(
            [
                27,
                29,
                3,
                12,
            ],
        ),
        Tet4Connectivity(
            [
                9,
                10,
                1,
                27,
            ],
        ),
        Tet4Connectivity(
            [
                9,
                1,
                0,
                27,
            ],
        ),
        Tet4Connectivity(
            [
                27,
                31,
                13,
                12,
            ],
        ),
        Tet4Connectivity(
            [
                27,
                31,
                10,
                13,
            ],
        ),
        Tet4Connectivity(
            [
                27,
                31,
                9,
                10,
            ],
        ),
        Tet4Connectivity(
            [
                27,
                31,
                12,
                9,
            ],
        ),
        Tet4Connectivity(
            [
                0,
                1,
                4,
                27,
            ],
        ),
        Tet4Connectivity(
            [
                0,
                4,
                3,
                27,
            ],
        ),
        Tet4Connectivity(
            [
                11,
                14,
                2,
                28,
            ],
        ),
        Tet4Connectivity(
            [
                14,
                5,
                2,
                28,
            ],
        ),
        Tet4Connectivity(
            [
                28,
                30,
                5,
                4,
            ],
        ),
        Tet4Connectivity(
            [
                28,
                30,
                14,
                5,
            ],
        ),
        Tet4Connectivity(
            [
                28,
                30,
                13,
                14,
            ],
        ),
        Tet4Connectivity(
            [
                28,
                30,
                4,
                13,
            ],
        ),
        Tet4Connectivity(
            [
                10,
                11,
                1,
                28,
            ],
        ),
        Tet4Connectivity(
            [
                11,
                2,
                1,
                28,
            ],
        ),
        Tet4Connectivity(
            [
                28,
                32,
                14,
                13,
            ],
        ),
        Tet4Connectivity(
            [
                28,
                32,
                11,
                14,
            ],
        ),
        Tet4Connectivity(
            [
                28,
                32,
                10,
                11,
            ],
        ),
        Tet4Connectivity(
            [
                28,
                32,
                13,
                10,
            ],
        ),
        Tet4Connectivity(
            [
                1,
                2,
                4,
                28,
            ],
        ),
        Tet4Connectivity(
            [
                2,
                5,
                4,
                28,
            ],
        ),
        Tet4Connectivity(
            [
                29,
                30,
                16,
                13,
            ],
        ),
        Tet4Connectivity(
            [
                29,
                30,
                7,
                16,
            ],
        ),
        Tet4Connectivity(
            [
                29,
                30,
                4,
                7,
            ],
        ),
        Tet4Connectivity(
            [
                29,
                30,
                13,
                4,
            ],
        ),
        Tet4Connectivity(
            [
                3,
                6,
                12,
                29,
            ],
        ),
        Tet4Connectivity(
            [
                6,
                15,
                12,
                29,
            ],
        ),
        Tet4Connectivity(
            [
                6,
                7,
                15,
                29,
            ],
        ),
        Tet4Connectivity(
            [
                7,
                16,
                15,
                29,
            ],
        ),
        Tet4Connectivity(
            [
                29,
                33,
                16,
                15,
            ],
        ),
        Tet4Connectivity(
            [
                29,
                33,
                13,
                16,
            ],
        ),
        Tet4Connectivity(
            [
                29,
                33,
                12,
                13,
            ],
        ),
        Tet4Connectivity(
            [
                29,
                33,
                15,
                12,
            ],
        ),
        Tet4Connectivity(
            [
                3,
                4,
                6,
                29,
            ],
        ),
        Tet4Connectivity(
            [
                4,
                7,
                6,
                29,
            ],
        ),
        Tet4Connectivity(
            [
                14,
                17,
                8,
                30,
            ],
        ),
        Tet4Connectivity(
            [
                14,
                8,
                5,
                30,
            ],
        ),
        Tet4Connectivity(
            [
                7,
                8,
                17,
                30,
            ],
        ),
        Tet4Connectivity(
            [
                7,
                17,
                16,
                30,
            ],
        ),
        Tet4Connectivity(
            [
                30,
                34,
                17,
                16,
            ],
        ),
        Tet4Connectivity(
            [
                30,
                34,
                14,
                17,
            ],
        ),
        Tet4Connectivity(
            [
                30,
                34,
                13,
                14,
            ],
        ),
        Tet4Connectivity(
            [
                30,
                34,
                16,
                13,
            ],
        ),
        Tet4Connectivity(
            [
                4,
                5,
                8,
                30,
            ],
        ),
        Tet4Connectivity(
            [
                4,
                8,
                7,
                30,
            ],
        ),
        Tet4Connectivity(
            [
                31,
                32,
                22,
                19,
            ],
        ),
        Tet4Connectivity(
            [
                31,
                32,
                13,
                22,
            ],
        ),
        Tet4Connectivity(
            [
                31,
                32,
                10,
                13,
            ],
        ),
        Tet4Connectivity(
            [
                31,
                32,
                19,
                10,
            ],
        ),
        Tet4Connectivity(
            [
                9,
                12,
                18,
                31,
            ],
        ),
        Tet4Connectivity(
            [
                12,
                21,
                18,
                31,
            ],
        ),
        Tet4Connectivity(
            [
                31,
                33,
                13,
                12,
            ],
        ),
        Tet4Connectivity(
            [
                31,
                33,
                22,
                13,
            ],
        ),
        Tet4Connectivity(
            [
                31,
                33,
                21,
                22,
            ],
        ),
        Tet4Connectivity(
            [
                31,
                33,
                12,
                21,
            ],
        ),
        Tet4Connectivity(
            [
                18,
                19,
                9,
                31,
            ],
        ),
        Tet4Connectivity(
            [
                19,
                10,
                9,
                31,
            ],
        ),
        Tet4Connectivity(
            [
                21,
                22,
                18,
                31,
            ],
        ),
        Tet4Connectivity(
            [
                22,
                19,
                18,
                31,
            ],
        ),
        Tet4Connectivity(
            [
                20,
                23,
                14,
                32,
            ],
        ),
        Tet4Connectivity(
            [
                20,
                14,
                11,
                32,
            ],
        ),
        Tet4Connectivity(
            [
                32,
                34,
                14,
                13,
            ],
        ),
        Tet4Connectivity(
            [
                32,
                34,
                23,
                14,
            ],
        ),
        Tet4Connectivity(
            [
                32,
                34,
                22,
                23,
            ],
        ),
        Tet4Connectivity(
            [
                32,
                34,
                13,
                22,
            ],
        ),
        Tet4Connectivity(
            [
                19,
                20,
                11,
                32,
            ],
        ),
        Tet4Connectivity(
            [
                19,
                11,
                10,
                32,
            ],
        ),
        Tet4Connectivity(
            [
                22,
                23,
                20,
                32,
            ],
        ),
        Tet4Connectivity(
            [
                22,
                20,
                19,
                32,
            ],
        ),
        Tet4Connectivity(
            [
                33,
                34,
                25,
                22,
            ],
        ),
        Tet4Connectivity(
            [
                33,
                34,
                16,
                25,
            ],
        ),
        Tet4Connectivity(
            [
                33,
                34,
                13,
                16,
            ],
        ),
        Tet4Connectivity(
            [
                33,
                34,
                22,
                13,
            ],
        ),
        Tet4Connectivity(
            [
                12,
                15,
                24,
                33,
            ],
        ),
        Tet4Connectivity(
            [
                12,
                24,
                21,
                33,
            ],
        ),
        Tet4Connectivity(
            [
                15,
                16,
                25,
                33,
            ],
        ),
        Tet4Connectivity(
            [
                15,
                25,
                24,
                33,
            ],
        ),
        Tet4Connectivity(
            [
                24,
                25,
                22,
                33,
            ],
        ),
        Tet4Connectivity(
            [
                24,
                22,
                21,
                33,
            ],
        ),
        Tet4Connectivity(
            [
                23,
                26,
                14,
                34,
            ],
        ),
        Tet4Connectivity(
            [
                26,
                17,
                14,
                34,
            ],
        ),
        Tet4Connectivity(
            [
                16,
                17,
                25,
                34,
            ],
        ),
        Tet4Connectivity(
            [
                17,
                26,
                25,
                34,
            ],
        ),
        Tet4Connectivity(
            [
                25,
                26,
                22,
                34,
            ],
        ),
        Tet4Connectivity(
            [
                26,
                23,
                22,
                34,
            ],
        ),
    ],
}
